### PR TITLE
InstallFixes2

### PR DESCRIPTION
**Pull request by dschaper**
Continuation of Install Fixes - Branch 2

### CODE REVIEW

File: automated install/basic-install.sh [hunk 1]
```diff
@@ -319,53 +320,57 @@ It is also possible to use a DHCP reservation, but if you are going to do that,
...
+	static ip_address=$IPv4_address
 	static routers=$IPv4gw
 	static domain_name_servers=$IPv4gw" | tee -a /etc/dhcpcd.conf >/dev/null
 }
 
 setStaticIPv4() {
+  local IFCFG_FILE
+  local IPADDR
+  local CIDR
 	if [[ -f /etc/dhcpcd.conf ]];then
```
Comment (dschaper) [inline automated install/basic-install.sh]: This will always exist. We are installing `dhcpcd5` as an installer dependency, and that file is automatically generated. So every Debian and derivative will have this file no matter what based on current install. Need ideas on how to handle this situation. Take `dhcpcd5` out of the install dependencies, but then how should we proceed?

Comment (dschaper) [inline automated install/basic-install.sh]:
```
pi@raspberrypi:/etc/pihole $ cat /etc/network/interfaces
# interfaces(5) file used by ifup(8) and ifdown(8)

# Please note that this file is written to be used with dhcpcd
# For static IP, consult /etc/dhcpcd.conf and 'man dhcpcd.conf'
```
This is the header for `/etc/network/interfaces` on Raspbian, it does not exist in other distributions in that file.

Comment (dschaper) [inline automated install/basic-install.sh]: And lastly, this is os-release
```
pi@raspberrypi:/etc/pihole $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 8 (jessie)"
NAME="Raspbian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=raspbian
ID_LIKE=debian
```

File: automated install/basic-install.sh [hunk 2]
```diff
@@ -319,53 +320,57 @@ It is also possible to use a DHCP reservation, but if you are going to do that,
...
+			echo "DEVICE=$piholeInterface"
+			echo "BOOTPROTO=none"
+			echo "ONBOOT=yes"
+			echo "IPADDR=$IPADDR"
+			echo "PREFIX=$CIDR"
+			echo "GATEWAY=$IPv4gw"
+			echo "DNS1=$piholeDNS1"
+			echo "DNS2=$piholeDNS2"
+			echo "USERCTL=no"
+			}>> "${IFCFG_FILE}"
```
Comment (diginc) [inline automated install/basic-install.sh]: HEREDOC this block? Just did this my self and almost used the style you implemented but switched to HEREDOC https://github.com/diginc/docker-pi-hole/pull/58/files#diff-f4647fb585ba39c152977803e97b4520R55

Like this?  `cat <<- EOF > $FILE` the dash after << gets rid of the whitespace but it only supports tabs...I personally was willing to trade echo and quotes for all of the tabs?

https://github.com/diginc/docker-pi-hole/pull/58/files#diff-f4647fb585ba39c152977803e97b4520R55

Comment (dschaper) [inline automated install/basic-install.sh]: Yeah, it was either heredoc or grouping and output. Didn't see the benefit of one over the other, actually.

File: automated install/basic-install.sh [hunk 3]
```diff
@@ -968,7 +979,7 @@ fi
 
 echo ":::"
 echo "::: The install log is located at: /etc/pihole/install.log"
-echo "::: View the web interface at http://pi.hole/admin or http://${IPv4addr%/*}/admin"
+echo "::: View the web interface at http://pi.hole/admin or http://${IPv4_address%/*}/admin"
 }
 
 main "$@"
```
Comment (diginc) [inline automated install/basic-install.sh]: Could we get a dry-run that can prevent execution of this to allow just sourcing the functions and running them one by one?

This would allow unit tests to be written for the functions.  Though the test framework could also just comment this line out it may help with manual testing too, to not run everything at once.

**Comment by PromoFaux** [comment]
Approved.

Tested on:
- [x] Ubuntu 16.05 DO Droplet
- [ ] Debian 8.6  DO Droplet (Some issues, but we know about them)
- [x] Rasbian Jessie on Pi2
- [x] Debian 8.6 on a laptop

**Comment by dschaper** [comment]
Approved

Will add trap to next branch that will explain the Debian 8.6 DO issue, just need to instruct the user to try an install again, right now it drops to prompt without explanation.

**Comment by dschaper** [comment]
And `dhcpcd5` will always make that sentinel file exist, so that's on the list of things to solve in the next round.

**Comment by PromoFaux** [comment]
It's probably edge-case at this stage, mind. I wonder how many people are running it on a DO droplet anyway.

**Comment by dschaper** [comment]
Very few, but it's easier to test there than in a VirtualBox VM :) Need to set them up with snapshots.

File: automated install/basic-install.sh (hunk 3 above)
Comment (diginc) [inline automated install/basic-install.sh]: when dry run is set, do not run the main / anything. `if [[ -z "$DRYRUN" ]] ; main "$@ ; fi;` so it would always run main by default.

then if I wanted to write a test for one function I could do something like:
```
DRYRUN=TRUE . automated install/basic-install.sh;
useIPv6dialog
assert $IPv6addr == 1.2.3.4
```

Comment (dschaper) [inline automated install/basic-install.sh]: Warner had something he was working on with an undocumented flag that would pull the variables from a file and run non-interactive. But I understand what you mean now. Instead of triggering `main()` on load with the flag set it would allow for everything to load up and prep for main but then you could unit/nosetest the functions... I'll look at that and see how to get that for you.

**Comment by PromoFaux** [comment]
have reset my pullApprove to Pending as this branch breaks Centos installs!

**Comment by dschaper** [comment]
Can you do that? I'll just respect that it should be in hold status until all agree.

File: gravity.sh [hunk 4]
```diff
@@ -254,10 +254,10 @@ function gravity_hostFormat() {
 		echo "::: Error: Unable to determine fully qualified domain name of host"
 	fi
 	# If there is a value in the $piholeIPv6, then IPv6 will be used, so the awk command modified to create a line for both protocols
-	if [[ -n ${piholeIPv6} ]];then
```
Comment (diginc) [inline gravity.sh]: quotes around variable?

Comment (dschaper) [inline gravity.sh]: I'm working mostly on `basic-install.sh` haven't gotten to much into `gravity.sh`

Comment (diginc) [inline gravity.sh]: I think -n and -z require quotes to work properly:

```
[ -z $asdf ] && echo hit
hit
[ -z "$asdf" ] && echo hit
hit
[ -n $asdf ] && echo hit
hit
[ -n "$asdf" ] && echo hit
```

The 2nd to last one shouldn't have echo'd hit.

File: gravity.sh [hunk 5]
```diff
@@ -254,10 +254,10 @@ function gravity_hostFormat() {
...
+	if [[ -n $"{IPv6_address}" ]];then
```
Comment (diginc) [inline gravity.sh]: `"${IPv6_address}"`

Comment (PromoFaux) [inline gravity.sh]: I'm ashamed to say it took me a full 30 seconds of staring at that trying to work out what you were pointing out. Good catch!

**Comment by dschaper** [comment]
@bcambl Can you run this branch through your CentOS and see how it works for you? It's not `set -e` so any errors will ditch out the script and I'm trying to track down all the logic places where we rely on return values for flow control. Also, with DO, I'm finding that you can't use the automagically selected IPv4 address because it grabs the 10 net address and that makes Network Manager really mad, you have to type in the public IP address in the choices box for it to pass that section. (Along with the fact that SEL is disabled in DO so had to code around that and not throw a shoe when the firewall and SEL tried to function. SEL will install the policies, but they are not in effect. Firewall just bypasses any entries since it's non-op)... This is a major change in the install and we don't want to break CentOS.

**Comment by bcambl** [comment]
```
[root@test automated install]# ./basic-install.sh 
:::
::: You are root.
::: Verifying free disk space...
:::
::: yum update has not been run today. Running now... done!
:::
::: Checking yum for upgraded packages.... done!
:::
::: There are 2 updates available for your system!
::: We recommend you run 'yum update -y' after installing Pi-Hole! 
:::
:::    Checking for iproute... installed!
:::    Checking for net-tools... installed!
:::    Checking for procps-ng... installed!
:::    Checking for newt... installed!
:::    Checking for git... installed!
:::    Checking for epel-release... installed!
:::    Checking for bind-utils... installed!
:::    Checking for bc... installed!
:::    Checking for dnsmasq... installed!
:::    Checking for lighttpd... installed!
:::    Checking for lighttpd-fastcgi... installed!
:::    Checking for php-common... installed!
:::    Checking for php-cli... installed!
:::    Checking for php... installed!
:::    Checking for curl... installed!
:::    Checking for unzip... installed!
:::    Checking for wget... installed!
:::    Checking for findutils... installed!
:::    Checking for cronie... installed!
:::    Checking for sudo... installed!
:::    Checking for nmap-ncat... installed!
:::
::: Checking for existing repository...
:::    Checking /etc/.pihole is a repo... OK!
:::     Updating repo in /etc/.pihole... done!
:::
::: Checking for existing repository...
:::    Checking /var/www/html/admin is a repo...:::    Cloning https://github.com/pi-hole/AdminLTE.git into /var/www/html/admin... done!
:::
::: Stopping dnsmasq service... done.
:::
::: Stopping lighttpd service... done.
::: Using interface: ens33
::: Warning: Unable to locate configuration file to set static IPv4 address!
[root@test automated install]#
```

**Comment by dschaper** [comment]
You're going to hate me, but can you `bash -x basic-install.sh`?

**Comment by dschaper** [comment]
There's something throwing the `set -e` and I don't have the trap set up yet to report to the user exactly what is going wrong...

**Comment by bcambl** [comment]
Ignore the previous comment. Turns out my test server network configuration is an edge case. My `ifcfg-` file name does not match my device name. This caused `-f /etc/sysconfig/network-scripts/ifcfg-<device>` to fail. I touched the file and then everything went well.

I have a few things to add here if possible.. mind if I send a couple pull requests?

**Comment by dschaper** [comment]
You are definitely more than welcome to PR what ever you'd like to contribute. We wouldn't have CentOS without your help!

**Comment by bcambl** [comment]
Ok, I will test my changes within the next days and submit a couple PR's to this branch.

`updatesToInstall=$(eval "${PKG_COUNT}")` seems to be failing now.

```
+ echo -n '::: Checking dnf for upgraded packages....'
::: Checking dnf for upgraded packages....++ eval dnf check-update '|' grep -v '^Last' '|' grep -c '^[a-zA-Z0-9]'
+++ dnf check-update
+++ grep -v '^Last'
+++ grep -c '^[a-zA-Z0-9]'
+ updatesToInstall=0
[root@localhost .pihole]# 
```

This works for me:
```
if [[ $(eval "${PKG_COUNT}") -eq "0" ]]; then
```

But then you will lose your package count.  
Also, This is also the only `eval` in the installer. We should try and get rid of it anyways.

**Comment by dschaper** [comment]
It may be the `grep -c`. If that evaluates to 0, then it will return an exit code of 1 and that will trip set -e. Check the package count for Debian and I think I left a note in there how to code around that one gotcha. It only happens when you have 0 packages to upgrade, any other value will pass without issue.

**Comment by PromoFaux** [comment]
Approved, in it's current state, having re-tested on the following distros
- [x] Ubuntu 16.05 DO Droplet
- [x] Debian 8.6  DO Droplet (Some issues, but we know about them, requires a second run of the script)
- [x] Rasbian Jessie on Pi2
- [x] Debian 8.6 on a laptop
- [x] Centos 7.2 DO
- [x] Centos 6.8 DO

**Comment by dschaper** [comment]
Okay, can we get this to merge to development and I'll continue to clean things up. I'd like to get some users to beta and doing that off devel may be easier.

**Comment by dschaper** [comment]
:+1:

**Comment by jacobsalmela** [comment]
Engage

**Comment by dschaper** [comment]
Going to try an experiment and see if we can keep this branch open and still push to it, and keep it so we can merge to development again without losing the history and conversation.

**Comment by iannao** [comment]
Hi Dschaper, i'm not able to make question in the other thread, can you please tell me how did you fix the issue? we are having the same problem in our app.

Thank you
Best Regards
Riccardo

**Comment by dschaper** [comment]
Please open a new thread, do not re-use existing threads for other purposes. In 30 minutes I will delete both of these comments from this thread as we need it for tracking other issues. Thank you for your understanding.
